Formattng all Python files with Black (+ Enforce)

TeddyAndrieux opened the pull request:
**Component**:

'formatting', 'python'

**Context**: 

#2650 

**Summary**:

- Format all Python files using black
- Check for "formatting" as part of python linting check
- Add some pre-commit hooks

---

Fixes: #2650 

---

Wait for #3066 to be merged

TeddyAndrieux committed:
tox: Add formatting for all Python files

Add a tox entry to format all Python files using black and run this
tox command with doit. Also add a check about format as part of
lint-python

- Set the 'max-line-length' to 88 in pylintrc since it's the default for
  Black
- Do not longer ignore 'multiple-spaces-before-operator' in Salt
  pylintrc since we use Black to format all the code
- Disable error 'whitespace-before-colon' in pylintrc since it's a known
  issue between Flake8 and Black
  (Sees: https://black.readthedocs.io/en/stable/the_black_code_style.html#slices)

Refs: #2650

TeddyAndrieux committed:
buildchain: Define a variable for default args of Skope copy

Add a constant variable for defaults arguments of Skope copy command so
that we avoid duplicate code

Fix lint error
```
R:  1, 0: Similar lines in 2 files
==buildchain.targets.local_image:200
==buildchain.targets.remote_image:123
            config.ExtCommand.SKOPEO.value,
            "--override-os",
            "linux",
            "--insecure-policy",
            "copy",
            "--format",
            "v2s2", (duplicate-code)
```

TeddyAndrieux committed:
buildchain: Use `SRC_DEB_ROOT` instead of `ROOT` path

Instead of computing again the path to `SRC_DEB_ROOT` in packaging
buildchain module and repository target module, use directly
a new `SRC_DEB_ROOT` constant

Fix lint error
```
R:  1, 0: Similar lines in 2 files
==buildchain.packaging:313
==buildchain.targets.repository:375
        mounts = [
            utils.bind_ro_mount(
                source=constants.ROOT
                / "packages"
                / "debian"
                / "common" (duplicate-code)
```

TeddyAndrieux committed:
tox: Add pre-commit config for Python formatting

Add a pre-commit config file that format Python files using black and
also add a tox entry to run this pre-commit

Move Python format from tox to pre-commit and update doit

TeddyAndrieux committed:
tox: Move Python linting from tox to pre-commit hook

Add a pre-commit hook in the pre-commit config file about Python linting
using Pylint. Remove python linting from tox and update doit

TeddyAndrieux committed:
tox: Move Python typing from tox to pre-commit hook

Add a pre-commit hook in the pre-commit config file about Python typing
check using mypy. Remove this entry from tox

TeddyAndrieux committed:
tox: Move Salt SLS lint from tox to pre-commit hook

Add a pre-commit hook in the pre-commit config file about Salt SLS
linting using salt-lint. Remove this entry from tox and update doit

NicolasT commented on .pre-commit-config.yaml:
```diff
@@ -0,0 +1,51 @@
+      - id: black
+        name: Formatting Python
+
+  - repo: https://github.com/pycqa/pylint
+    rev: pylint-2.6.0
+    hooks:
+      - id: pylint
+        name: Lint Python (non-salt)
+        # Linting for non-salt files (also exclude tests files)
+        additional_dependencies: ['pyenchant~=2.0', 'doit~=0.32.0']
+        exclude: >
+          (?x)^(
+              salt/_.*|
+              salt/tests/.*|
+              tests/.*
+          )$
+      - id: pylint
+        name: Lint Salt
+        # Linting for salt files only in directory "salt/_<module>/<module_name>.py"
+        files: salt/_.*/.*\.py
+        additional_dependencies: ['saltpylint==2020.9.28']
+        args:
+          - --rcfile=salt/.pylintrc
+          - --ignore=metalk8s_package_manager_apt.py
+
+  - repo: https://github.com/pre-commit/mirrors-mypy
+    rev: v0.740
+    hooks:
+      - id: mypy
+        name: Python typing
+        # Only check typing for typed files
+        files: >
+          (?x)^(
+              buildchain/dodo.py|
+              buildchain/buildchain/.*|
+              packages/debian/common/download_packages.py
+          )$
+        args:
+          - --strict
```
Not super fond of this. Now this needs to be kept aligned between here and `tox.ini` (which should be the canonical source).

TeddyAndrieux replied:
I agree, for the moment that commands are really simple so... shouldn't be really hard to maintain but indeed I would prefer a single entry.
We could imagine using pre-commit for linting, formatting, unit tests instead of tox and in tox we only do a pre-commit run

Or we remove this pre-commit hook :shrug:

TeddyAndrieux replied:
Any opinion on this one @scality/metalk8s ?

gdemonet replied:
Couldn't we call `tox` from a pre-commit hook? And we could also render the hook to the right place using `doit`, no need for another tool to `pre-commit install` (personally just discovered it :sweat_smile:).

TeddyAndrieux replied:
I didn't found any existing hook for tox but it shouldn't be hard to create one but even if we create a hook for tox we need to rewrite a bit our tox so that we are able to run only on one specific file.



> And we could also render the hook to the right place using `doit`, no need for another tool to `pre-commit install`

WDYM ?

gdemonet replied:
> we need to rewrite a bit our tox so that we are able to run only on one specific file.

Likely a good idea, yes.

> WDYM ?

That instead of using this `pre-commit` tool to install a magic `.git/hooks/pre-commit` script which parses this `.pre-commit-config.yaml`, just put our own `.git/hooks/pre-commit` script and have it do whatever we need. Unless you really think we should adopt this tool as a standard that we expect developers to have pre-installed (otherwise they can't really enable this hook).

TeddyAndrieux replied:
For me `pre-commit` is kind of common in the Python world and it's not because we have a pre-commit config, that all developer needs to have pre-commit hook installed it's just to help the developer if they want to.
Also, a dev can just run `tox -e pre-commit` to run this pre-commit "hook" manually and do not need to install pre-commit directly. Or just wait for a build that will basically check that linting&co is good as well.

IMHO we should define the lining stuff directly in pre-commit config and in tox we just call pre-commit on all files, from what I saw on various python project it's something fairly common

NicolasT commented on .pylintrc:
```diff
@@ -237,7 +237,7 @@ indent-after-paren=4
 indent-string='    '
 
-max-line-length=80    # Maximum number of characters on a single line.
+max-line-length=88    # Maximum number of characters on a single line.
```
:thinking:

TeddyAndrieux replied:
Just sync with black defaults, but if we want to keep 80 I could also change the black value

NicolasT commented on buildchain/buildchain/builder.py:
```diff
@@ -32,96 +30,113 @@ def _builder_image(
-                'yum_repositories'/'kubernetes.repo',
-            constants.ROOT/'packages'/'redhat'/'common'/
-                'yum_repositories'/'saltstack.repo',
+            constants.ROOT
+            / "packages"
+            / "redhat"
+            / "common"
+            / "yum_repositories"
+            / "kubernetes.repo",
```
:disappointed:

gdemonet replied:
Should likely be rewritten as:
```python
constants.ROOT.joinpath(
    "packages/redhat/common",
	"yum_repositories/kubernetes.repo",
)
```

Or:
```python
REDHAT_REPOS = constants.ROOT.joinpath(
    "packages/redhat/common/yum_repositories",
)

# ...
file_dep = [
    REDHAT_REPOS / "kubernetes.repo",
    REDHAT_REPOS / "saltstack.repo",
]
```

TeddyAndrieux replied:
```python
file_dep=[
    constants.ROOT
    / "packages/redhat/common/yum_repositories"
    / "kubernetes.repo",
    constants.ROOT
    / "packages/redhat/common/yum_repositories"
    / "saltstack.repo",
]
```

~Good also no ?~

Edit: No ok with "REDHAT_REPOS" variable looks better especially since we do the same for redhat 8

NicolasT reviewed (commented):
Impossible to review this.

Can you at least make sure the commit which reformats everything and the kitchen-sync is added to the metadata Git then uses to 'skip' said commit in `git blame` and others?

TeddyAndrieux committed:
git: Add a git blame ignore revs file for formatting commit

Add a file to list all the revs to ignore by git blame with
`git blame --ignore-revs-file .git-blame-ignore-revs` or directly in the
Git config.
We add this file to ignore the formatting commit when we use git blame

TeddyAndrieux commented:
> Impossible to review this.

Not sure what I can do about it

> Can you at least make sure the commit which reformats everything and the kitchen-sync is added to the metadata Git then uses to 'skip' said commit in `git blame` and others?

Don't know it was possible, I just created a `.git-blame-ignore-revs` file that can be used with `git blame --ignore-revs-file .git-blame-ignore-revs`, that's what you mean right ?

gdemonet commented on buildchain/buildchain/targets/local_image.py:
```diff
@@ -197,16 +197,8 @@ def _do_build(self) -> List[types.Action]:
-            "v2s2",
-            "--dest-compress",
-        ]
+        cmd = list(constants.SKOPEO_COPY_DEFAULT_ARGS)
+        cmd.append("--dest-compress")
```
Python 3.5+ shorthand:

```python
cmd = [*constants.SKOPEO_COPY_DEFAULT_ARGS, "--dest-compress"]
```

gdemonet reviewed (commented):
PR looks good to me, but it will require some explanations / docs for onboarding the team to the new tools introduced:

- how to setup the pre-commit hooks locally + which dependencies should be installed, or how to run them using tox
- make sure to setup your IDE with black formatting + pylint / salt-lint / mypy validation

Maybe you could add a small section to developer docs for this?

bert-e commented:
# Hello teddyandrieux,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Branches have diverged



This pull request's source branch `improvement/format-python-files` has diverged from
`development/2.8` by more than 50 commits.

To avoid any integration risks, please re-synchronize them using one of the
following solutions:

* Merge `origin/development/2.8` into `improvement/format-python-files`
* Rebase `improvement/format-python-files` onto `origin/development/2.8`

Note: If you choose to rebase, you may have to ask me to rebuild
integration branches using the `reset` command.

TeddyAndrieux committed:
Format all Python code with black

Fixes: #2650

TeddyAndrieux committed:
buildchain: Lint remove uneeded disable of "line-too-long"

Remove uneeded disable of "line-too-long" from salt_tree
```
I:502, 0: Useless suppression of 'line-too-long' (useless-suppression)
I:689, 0: Useless suppression of 'line-too-long' (useless-suppression)
```

TeddyAndrieux committed:
buildchain: Lint fix targets.package "implicit-str-concat"

Do not implicit concat string, instead just create on string
```
W:158, 0: Implicit string concatenation found in assignment (implicit-str-concat)
```

TeddyAndrieux committed:
buildchain: Define a variable for yum repositories directory

Add a variable Path in buildchain for the root directory that containts
all the yum repositories definition

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @thomasdanan

* @alexandre-allard-scality

* @jbwatenbergscality

* @gdemonet

* @lucieleonard

* @chengyanjin

* @teddyandrieux

* @nicolast

TeddyAndrieux committed:
tox: Move Yaml lint from tox to pre-commit hook

Add a pre-commit hook in the pre-commit config file about Yaml linting
using yamllint. Remove this entry from tox and update doit

TeddyAndrieux committed:
build: Remove `requirements-dev` file

Remove the `requirements-dev` file as this one is no longer needed,
dependencies are handled by the pre-commit config file

TeddyAndrieux committed:
pre-commit: Bump salt-lint to 0.5.2

Bump salt-lint to version 0.5.2 and remove no longer needed skip (as the
issue is fixed in 0.5.2)

TeddyAndrieux committed:
docs: Add a "pre-commit hooks" section in developer guide

Add a paragraph in developer guide documentation about this `pre-commit`
hooks

alexandre-allard commented on docs/developer/development/commit.rst:
```diff
@@ -1,6 +1,25 @@
 Commit Best Practices
 ---------------------
 
+Pre-commit hook
```
```suggestion
Pre-commit hooks
```

alexandre-allard commented on docs/developer/development/commit.rst:
```diff
@@ -1,6 +1,25 @@
+Pre-commit hook
+~~~~~~~~~~~~~~~
+
+Some pre-commit hook are define to do some linting check and also to format all
```
```suggestion
Some pre-commit hooks are defined to do some linting checks and also to format all
```

alexandre-allard commented on docs/developer/development/commit.rst:
```diff
@@ -1,6 +1,25 @@
+Some pre-commit hook are define to do some linting check and also to format all
+Python code automatically.
+
+Those checks are also done in the CI so make sure it succeed.
```
```suggestion
Those checks are also run in the CI pre-merge test suite to enforce code linting.
```

alexandre-allard commented on docs/developer/development/commit.rst:
```diff
@@ -1,6 +1,25 @@
+
+Those checks are also done in the CI so make sure it succeed.
+
+To run pre-commit automatically when you commit with git just install
+pre-commit ``pip install pre-commit``, and then install the hook
+``pre-commit install``.
```
```suggestion
To enable pre-commit hook to run automatically when committing, install it as follows:

.. code-block:: shell

   pip install pre-commit
   pre-commit install
```

alexandre-allard commented on docs/developer/development/commit.rst:
```diff
@@ -1,6 +1,25 @@
+You can skip this pre-commit hook on a specific commit
+``git commit --no-verify``.
+
+To run pre-commit manually just use tox directly ``tox -e pre-commit``.
+It's also possible to run only a specific hook (e.g. for pylint
+``tox -e pre-commit pylint``).
```
```suggestion
To run pre-commit manually, use tox:

.. code-block:: shell

   tox -e pre-commit

It is also possible to run only a specific hook (e.g. for pylint
``tox -e pre-commit pylint``).
```

alexandre-allard reviewed (commented):
LGTM

TeddyAndrieux commented:
/approve

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/5b3ca3a9b63f20b0a8a4fecbcd46ba5c322c860c) did not succeed in branch improvement/format-python-files.




*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.8`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.8`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`



Please check the status of the associated issue None.

Goodbye teddyandrieux.
